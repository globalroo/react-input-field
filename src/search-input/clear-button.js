import * as React from "react";
import PropTypes from "prop-types";
import "font-awesome/css/font-awesome.min.css";
import { styles, BUTTON_WIDTH } from "src/search-input/styles/styles";

export const ClearButton = ({ visible, onClick, hovered }) => (
  <button
    onClick={onClick}
    style={{
      width: BUTTON_WIDTH,
      border: 0,
      verticalAlign: "middle",
      opacity: visible ? 1 : 0
    }}
  >
    <span
      style={hovered ? styles.clearButtonIconHover : styles.clearButtonIcon}
      className="clear-button-icon fa-stack fa-lg"
    >
      <i
        style={{ opacity: hovered ? 0 : 1 }}
        className="fa fa-circle-o fa-stack-2x"
      />
      <i className="fa fa-times fa-stack-1x" />
    </span>
  </button>
);

ClearButton.propTypes = {
  hovered: PropTypes.bool,
  onClick: PropTypes.func,
  visible: PropTypes.bool
};
