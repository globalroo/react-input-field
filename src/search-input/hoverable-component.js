import * as React from "react";
import PropTypes from "prop-types";

/*
** Adds mouse in / out state indicator to any child components
** contained within.
*/
export class Hoverable extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			hovered: false
		};
	}
	setRef = node => (this.ref = node);
	handleMouseEnter = () => {
		this.setState({
			hovered: true
		});
	};
	handleMouseLeave = () => {
		this.setState({
			hovered: false
		});
	};
	componentDidMount() {
		this.mouseEnterListener = this.ref.addEventListener(
			"mouseenter",
			this.handleMouseEnter
		);
		this.mouseLeaveListener = this.ref.addEventListener(
			"mouseleave",
			this.handleMouseLeave
		);
	}
	componentWillUnmount() {
		this.ref.removeEventListener("mouseenter", this.mouseEnterListener);
		this.ref.removeEventListener("mouseleave", this.mouseLeaveListener);
	}
	render() {
		return (
			<div ref={this.setRef}>
				{React.cloneElement(this.props.children, this.state)}
			</div>
		);
	}
}

Hoverable.propTypes = {
	children: PropTypes.element.isRequired
};
