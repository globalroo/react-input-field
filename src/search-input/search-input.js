import * as React from "react";
import PropTypes from "prop-types";
import { assignStyles } from "src/utils/assign-styles";
import {
	BORDER_WIDTH,
	BUTTON_WIDTH,
	INPUT_LEFT_OFFSET,
	styles
} from "src/search-input/styles/styles";

import { Hoverable } from "src/search-input/hoverable-component";
import { ClearButton } from "src/search-input/clear-button";

export class SearchInput extends React.Component {
	constructor(props) {
		super(props);
		this.inputRef = null;
		this.borderRadius = Math.floor(this.props.height / 2);
		this.inputWidth =
			this.props.width - BORDER_WIDTH * 2 - BUTTON_WIDTH - INPUT_LEFT_OFFSET;
		this.state = {
			content: "",
			showButton: false
		};
	}

	setInputRef = node => (this.inputRef = node);
	setFocus = () => this.inputRef.focus();

	clearText = () => {
		const content = "";
		this.inputRef.value = content;
		this.setState({ content, showButton: false });
	};

	onChangeText = event => {
		const content = event.target.value;
		const showButton = content && content.length > 0 ? true : false;
		this.setState({ content, showButton });
	};

	componentDidMount() {
		this.props.whenAvailable(this.inputRef);
	}

	render() {
		return (
			<div
				onClick={this.setFocus}
				style={assignStyles([
					{
						width: this.props.width,
						height: this.props.height,
						borderRadius: this.borderRadius
					},
					styles.componentContainer
				])}
			>
				<input
					type="text"
					style={assignStyles([
						styles.inputStyles,
						{
							marginLeft: INPUT_LEFT_OFFSET,
							width: this.inputWidth
						}
					])}
					placeholder={this.props.placeholder}
					onChange={this.onChangeText}
					ref={node => (this.inputRef = node)}
				/>
				<Hoverable>
					<ClearButton
						visible={this.state.showButton}
						onClick={this.clearText} /*hovered injected from Hoverable*/
					/>
				</Hoverable>
			</div>
		);
	}
}

SearchInput.propTypes = {
	width: PropTypes.number,
	height: PropTypes.number,
	placeholder: PropTypes.string,
	whenAvailable: PropTypes.func
};
