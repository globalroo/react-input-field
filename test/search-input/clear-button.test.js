import React from "react";
import * as renderer from "react-test-renderer";
import { ClearButton } from "src/search-input/clear-button";

describe("Search Input tests", () => {
	test("Renders the component as expected in hovered state", () => {
		const spy = jest.fn();
		const component = renderer.create(
			<ClearButton visible={true} hovered={true} onClick={spy} width={50} />
		);
		expect(component.toJSON()).toMatchSnapshot();
	});
	test("Renders the component as expected in unhovered state", () => {
		const spy = jest.fn();
		const component = renderer.create(
			<ClearButton visible={true} hovered={false} onClick={spy} width={50} />
		);
		expect(component.toJSON()).toMatchSnapshot();
	});
	test("Renders the component as expected in non-visible state", () => {
		const spy = jest.fn();
		const component = renderer.create(
			<ClearButton visible={false} hovered={false} onClick={spy} width={50} />
		);
		expect(component.toJSON()).toMatchSnapshot();
	});
});
// width: PropTypes.number,
// hovered: PropTypes.bool,
// onClick: PropTypes.func,
// visible: PropTypes.bool
