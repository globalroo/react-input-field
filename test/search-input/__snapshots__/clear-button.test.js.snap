// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Input tests Renders the component as expected in hovered state 1`] = `
<button
  onClick={[Function]}
  style={
    Object {
      "border": 0,
      "opacity": 1,
      "verticalAlign": "middle",
      "width": 45,
    }
  }
>
  <span
    className="clear-button-icon fa-stack fa-lg"
    style={
      Object {
        "animation": "App-logo-spin infinite 10s linear",
        "color": "red",
        "fontSize": "20px",
      }
    }
  >
    <i
      className="fa fa-circle-o fa-stack-2x"
      style={
        Object {
          "opacity": 0,
        }
      }
    />
    <i
      className="fa fa-times fa-stack-1x"
    />
  </span>
</button>
`;

exports[`Search Input tests Renders the component as expected in non-visible state 1`] = `
<button
  onClick={[Function]}
  style={
    Object {
      "border": 0,
      "opacity": 0,
      "verticalAlign": "middle",
      "width": 45,
    }
  }
>
  <span
    className="clear-button-icon fa-stack fa-lg"
    style={
      Object {
        "color": "#61dafb",
        "fontSize": "14px",
      }
    }
  >
    <i
      className="fa fa-circle-o fa-stack-2x"
      style={
        Object {
          "opacity": 1,
        }
      }
    />
    <i
      className="fa fa-times fa-stack-1x"
    />
  </span>
</button>
`;

exports[`Search Input tests Renders the component as expected in unhovered state 1`] = `
<button
  onClick={[Function]}
  style={
    Object {
      "border": 0,
      "opacity": 1,
      "verticalAlign": "middle",
      "width": 45,
    }
  }
>
  <span
    className="clear-button-icon fa-stack fa-lg"
    style={
      Object {
        "color": "#61dafb",
        "fontSize": "14px",
      }
    }
  >
    <i
      className="fa fa-circle-o fa-stack-2x"
      style={
        Object {
          "opacity": 1,
        }
      }
    />
    <i
      className="fa fa-times fa-stack-1x"
    />
  </span>
</button>
`;
